### ISSUE TITLE
Allow users to log in using either their email address or username

### ISSUE THREAD

**Issue by emmajclegg** [younginnovations/iatipublisher#1402]
Users currently need to log in to Publisher with the username they created on sign up and cannot use their email address in place of this. This is not obvious as the error message given after trying to log in with an email address (screenshot below) does not mention needing the username.

This creates support queries as users are more likely to forget their username, which they can't recover through the current password reset link.

Can we assess what work would be required to give users the flexibility to log in to Publisher using either their username or email address?

![image](https://github.com/younginnovations/iatipublisher/assets/137316374/bfbccafc-349f-448c-8826-53e98cefe1fc)

**Comment by emmajclegg** [comment]
@praweshsth - see Rob's question above. 

We're starting to think about this in the context of the IATI Registry CKAN API change Rob mentioned, and working towards single sign-on across IATI tools.

**Comment by praweshsth** [comment]
For now, let's not remove the username from the UI completely. we can change the login using email address instead of username.

**Comment by emmajclegg** [comment]
@praweshsth - can you give us more detail on what usernames are currently used for? (Rob's question). This is feeding into wider authentication work across IATI tools as mentioned, so it's important we understand the dependencies.

Also - I think login via email address OR username is the most flexible here. But useful to understand the context around usernames first.

**Comment by emmajclegg** [comment]
(following on from question above)

To minimise disruption to existing users, it seems like allowing users to login with either their username or email address offers most flexibility in the short term.

For new users, however - if usernames are only collected by IATI Publisher to create/link to an IATI Registry user account, could we look to autogenerate these for new registrations perhaps? (i.e. instead of users having to enter their own username in the registration step below). In future, we don't see a role for usernames across IATI tools and will start to phase these out.

![image](https://github.com/younginnovations/iatipublisher/assets/137316374/0b8ce4da-7b4d-4868-9409-53f5fea5b40c)

**Comment by praweshsth** [comment]
# Approach for username autogeneration

- Username will be autogenerated using the full name with underscore for spaces and all in lowercase
- while generating the username, if is not unique, a random number of 2 digits will be added after the username followed by a underscore.

**Comment by PG-Momik** [comment]
@emmajclegg 

> praweshsth - can you give us more detail on what usernames are currently used for? (Rob's question). This is feeding into wider authentication work across IATI tools as mentioned, so it's important we understand the dependencies

> ...usernames are only collected by IATI Publisher to create/link to an IATI Registry user account..

Yes `username` is primarily used for authentication in Publisher. 

> [!NOTE]
> - Username & Password used when creating an organisation via 'I am new to IATI' registration option lets, users login to registry using those Username and password.
> - Username & Password used when creating an organisation via 'My organisation has registered with IATI' registration option cannot be used to login into registry.
> - We can only login using Username & Password in IATI registry. No email & password login.

We can remove login via Username & password when needed.  

We've enabled login via Email & Password and Username & Password. (Currently being tested)

---
 **cc** : @praweshsth

**Comment by sarinasindurakar** [comment]
@PG-Momik 
1. The user are not able to login with both  username and email.
![image](https://github.com/younginnovations/iatipublisher/assets/65848732/f7c3b627-9907-4535-b050-1b518e56e92c)

**Comment by sarinasindurakar** [comment]
@emmajclegg 
I attempted to replicate the initial issue **(error: 500)** you mentioned by registering the organization, but we were unable to replicate. Here I have attached a video demonstrating the steps I took to replicate and register the organization.

[Untitled_ Apr 03,2024 2_02 PM.webm](https://github.com/younginnovations/iatipublisher/assets/65848732/b5ff8183-1719-40a5-b040-fdde7e520f18)

Could you please provide a video illustrating how you encountered these issues? 
Additionally, could you clarify if this issue occurred only once or if it persists consistently? 
Other question regarding the error:500 , how long was the form idle i.e how long was the wait time between clicking 'Next Step' on step 3 and 4.

**Comment by emmajclegg** [comment]
Hi @sarinasindurakar - I haven't been able to recreate the registration issue either, so am not sure what caused it. I've been able to register a test organisation successfully since. I'll let you know if I see the error message again but assume it's no longer a problem.

For this issue - allowing users to login with their email or username - I've tested with a couple of user accounts and all looks to be working as expected. So, happy for this fix to be deployed. Thanks
